File: pandas/general_query_15.py
#!/usr/bin/python3

"""General query 15

Student Mobility

Javier Fernández-Bravo Peñuela
Universitat Politécnica de València

Total exchange students having been taught per teaching language.

pandas - general_query_15.py
"""

from pandas_job.pandas_task import compute_task
from domain.student import Student

from sys import argv

def teaching_language(students_df):
    students_df['LANGUAGETAUGHT'] = students_df['LANGUAGETAUGHT']\
        .apply(lambda lt: lt.upper() if len(lt) > 0 else None)
    students_df = students_df.dropna(subset=['LANGUAGETAUGHT'])
    return students_df[['LANGUAGETAUGHT']].groupby(['LANGUAGETAUGHT']).size()

# python3 general_query_15.py ../data/Student_Mobility.csv [out/pandas15-output]
if __name__ == "__main__":
    compute_task(teaching_language, Student, argv[1:])

File: hadoop/src/main/java/systematicchaos/studentmobility/GeneralQuery15.java
/**
 * Student Mobility
 * 
 * Javier Fernández-Bravo Peñuela
 * Universitat Politècnica de València
 * 
 * General query 15: Total exchange students having been taught per teaching language.
 * 
 * hadoop - GeneralQuery15.java
 */

package systematicchaos.studentmobility;

import java.util.Iterator;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;

import systematicchaos.studentmobility.domain.Student;
import systematicchaos.studentmobility.hadoop.HadoopJobLauncher;
import systematicchaos.studentmobility.hadoop.HadoopMapper;
import systematicchaos.studentmobility.hadoop.HadoopReducer;
import systematicchaos.studentmobility.util.KeyValuePair;

public class GeneralQuery15 {
	
	public static class TeachingLanguageMapper extends HadoopMapper<Text, LongWritable> {
		
		private static final LongWritable ONE = new LongWritable(1l);
		private Text teachingLanguage = new Text();
		
		private static final String HEADER = "HOMEINSTITUTION";
		
		@Override
		public KeyValuePair<Text, LongWritable> map(String token) {
			KeyValuePair<Text, LongWritable> mapResult = null;
			if (!token.startsWith(HEADER)) {
				Student student = Student.fromString(token);
				String languageTaught = student.getLanguageTaught();
				if (languageTaught != null && languageTaught.length() > 0) {
					teachingLanguage.set(languageTaught.toUpperCase());
					mapResult = new KeyValuePair<>(teachingLanguage, ONE);
				}
			}
			return mapResult;
		}
	}
	
	public static class TeachingLanguageReducer extends HadoopReducer<Text, LongWritable> {
		
		private LongWritable reduceResult = new LongWritable();
		
		@Override
		public LongWritable reduce(Text key, Iterator<LongWritable> values) {
			long count = 0l;
			while (values.hasNext()) {
				count += values.next().get();
			}
			reduceResult.set(count);
			return reduceResult;
		}
	}
	
	// hadoop jar target/hadoop-1.0.0.jar systematicchaos.studentmobility.GeneralQuery15 ../data/Student_Mobility.csv out/hadoop15-output
	public static void main(String[] args) throws Exception {
		final int exitCode = HadoopJobLauncher.launchHadoopJob(
				GeneralQuery15.class.getName(),
				TeachingLanguageMapper.class, TeachingLanguageReducer.class,
				Text.class, LongWritable.class, args);
		System.exit(exitCode);
	}
}
